packages/core/src/routes/admin-user.ts

```diff
@@ -0,0 +1,43 @@
+import { userInfoSelectFields } from '@logto/schemas';
+import pick from 'lodash.pick';
+import { InvalidInputError } from 'slonik';
+import { object, string } from 'zod';
+
+import koaGuard from '@/middleware/koa-guard';
+import { findRolesByRoleName } from '@/queries/roles';
+import { findUserById, updateUserById } from '@/queries/user';
+
+import { AuthedRouter } from './types';
+
+export default function adminUserRoutes<T extends AuthedRouter>(router: T) {
+  router.patch(
+    '/users/:userId/roles',
+    koaGuard({
+      params: object({ userId: string().min(1) }),
+      body: object({ roleNames: string().array().nullable() }),
+    }),
+    async (ctx, next) => {
+      const {
+        params: { userId },
+        body: { roleNames },
+      } = ctx.guard;
+
+      await findUserById(userId);
+
+      // Temp solution to validate the existence of input roleNames
+      if (roleNames?.length) {
```

Comment:
this solution is good i think

Reply:
I have a research issue on how to add the foreign key constraints in a jsonb field. Feels like that should be a DB level validation, not API level.